plugin/src/main/java/org/opensearch/sql/plugin/utils/XContentParserUtils.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.sql.plugin.utils;
+
+import static org.opensearch.common.xcontent.XContentParserUtils.ensureExpectedToken;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.opensearch.common.xcontent.XContentFactory;
+import org.opensearch.common.xcontent.XContentType;
+import org.opensearch.core.xcontent.DeprecationHandler;
+import org.opensearch.core.xcontent.NamedXContentRegistry;
+import org.opensearch.core.xcontent.XContentBuilder;
+import org.opensearch.core.xcontent.XContentParser;
+import org.opensearch.sql.datasource.model.DataSourceMetadata;
+import org.opensearch.sql.datasource.model.DataSourceType;
+
+public class XContentParserUtils {
+  public static final String NAME_FIELD = "name";
+  public static final String CONNECTOR_FIELD = "connector";
+  public static final String PROPERTIES_FIELD = "properties";
+  public static final String ALLOWED_ROLES_FIELD = "allowedRoles";
+
+  /**
+   * Convert xcontent parser to DataSourceMetadata.
+   *
+   * @param parser parser.
+   * @return DataSourceMetadata {@link DataSourceMetadata}
+   * @throws IOException IOException.
+   */
+  public static DataSourceMetadata toDataSourceMetadata(XContentParser parser) throws IOException {
```

Comment:
I think we use ObjectMapper before? found any issue?

Reply:
It needs Privileged permissions for reflections, so I removed it. I have verified other plugins and everyone is using XContentParser which also helps in transport apis where clients can pass XContentBuilders.